File: 01_composition/01_1.hs
-- f::A->B
-- g::B->C
--
-- g_after_f :: A -> C
-- g_after_f a = g . f

double :: Int -> Int
double x = x * 2

addOne :: Int -> Int
addOne x = x + 1

-- 合成
doubleThenAdd :: Int -> Int
doubleThenAdd = addOne . double

main :: IO ()
main = do
  print (doubleThenAdd 3)   -- (3*2)+1 = 7
  print (doubleThenAdd 10)  -- (10*2)+1 = 21

File: 01_composition/01_1.rs
fn f(x: i32) -> i32 {
    x * 2
}

fn g(x: i32) -> i32 {
    x + 1
}

fn compose<A, B, C, F, G>(f: F, g: G) -> impl Fn(A) -> C
where
    F: Fn(A) -> B,
    G: Fn(B) -> C,
{
    move |x| g(f(x))
}

fn main() {
    let g_after_f = compose(f, g);

    println!("{}", g_after_f(3));   // 7
    println!("{}", g_after_f(10));  // 21
}
